File: bindings/kotlin/scp-kt/src/main/kotlin/works/limn/scp/Lifecycle.kt
// Lifecycle.kt — Kotlin SDK bridge lifecycle wrapper
//
// Wraps the UniFFI scp_suspend / scp_resume functions as Kotlin suspend
// helpers. Bindings are injected via LifecycleBindings so tests can stub
// the FFI layer.
//
// Provenance: scp-ffi-uniffi lib.rs (scp_suspend / scp_resume).

@file:Suppress("MatchingDeclarationName")

package works.limn.scp

import works.limn.scp.bridge.CoroutineBridge

/**
 * Bindings for the bridge lifecycle UniFFI exports.
 *
 * Both methods are blocking JNA calls into Rust and must be dispatched
 * on [kotlinx.coroutines.Dispatchers.IO].
 */
interface LifecycleBindings {
    /** Suspends the bridge instance (disconnects transport, sets suspended flag). */
    fun scpSuspend()

    /** Resumes a suspended bridge instance (clears suspended flag). */
    fun scpResume()
}

/**
 * Namespace object for default bindings; the generated UniFFI functions are
 * global, and this object lets SDK callers choose between the real bindings
 * and an injected stub for tests.
 */
object LifecycleBridge {
    /**
     * Default bindings — delegate to the UniFFI-generated `scpSuspend` /
     * `scpResume` top-level functions. The `internal/` package is generated
     * at build time (see `scripts/generate-uniffi-kotlin.sh`); tests
     * injecting a stub should supply their own [LifecycleBindings] instead.
     */
    val default: LifecycleBindings = object : LifecycleBindings {
        override fun scpSuspend() {
            uniffi.scp.scpSuspend()
        }

        override fun scpResume() {
            uniffi.scp.scpResume()
        }
    }
}

/**
 * Suspend the bridge instance for mobile/desktop backgrounding.
 *
 * Disconnects the transport (clearing the relay connection) and marks
 * the instance as suspended. Context state is preserved — the instance
 * remains alive but inactive. Transport-dependent operations will fail
 * until [resume] is called.
 *
 * After suspension, callers should call [resume] to re-activate and
 * then re-establish the relay connection via `transportConnect`.
 *
 * No-op if the bridge has not been initialized or has already shut down.
 *
 * @param bridge The coroutine bridge (used for dispatcher routing).
 * @param bindings Injectable bindings for testing; defaults to the
 *   UniFFI-generated functions.
 * @throws uniffi.scp.ScpException if transport cleanup fails.
 */
suspend fun suspend(
    bridge: CoroutineBridge,
    bindings: LifecycleBindings = LifecycleBridge.default,
) {
    bridge.ffiCall { bindings.scpSuspend() }
}

/**
 * Resume a suspended bridge instance.
 *
 * Clears the suspended flag so bridge operations can proceed. The caller
 * must re-establish the relay connection via `transportConnect` —
 * resume does not reconnect automatically.
 *
 * No-op if the bridge is not initialized.
 *
 * @param bridge The coroutine bridge (used for dispatcher routing).
 * @param bindings Injectable bindings for testing; defaults to the
 *   UniFFI-generated functions.
 * @throws uniffi.scp.ScpException if the bridge has been permanently shut down.
 */
suspend fun resume(
    bridge: CoroutineBridge,
    bindings: LifecycleBindings = LifecycleBridge.default,
) {
    bridge.ffiCall { bindings.scpResume() }
}

File: bindings/python/scp_sdk/lifecycle.py
"""Bridge lifecycle controls for the SCP Python SDK.

Exposes :func:`suspend` and :func:`resume` which disconnect the bridge
from its relay (preserving context state) and clear the suspended flag,
respectively.  Use when backgrounding a mobile/desktop app, then call
:func:`resume` plus :func:`scp_sdk.connect_relay` to rejoin.

Both functions delegate to methods on an explicit :class:`scp_sdk.SCP`
instance (``suspend`` / ``resume``). After #1549 Phase 4 PR 4 the
default-instance façade is gone — callers thread an explicit
:class:`scp_sdk.SCP` through every operation.
"""

from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING

from scp_sdk.errors import ContextError, TransportError

if TYPE_CHECKING:
    from scp_sdk.scp import SCP


def suspend(scp: SCP) -> None:
    """Suspend the given bridge instance for backgrounding.

    Disconnects the transport (clearing the relay connection) and marks
    the instance as suspended.  Context state is preserved — the
    instance remains alive but inactive.  Transport-dependent operations
    will fail until :func:`resume` is called.

    After suspension, callers should call :func:`resume` to re-activate
    and then re-establish the relay connection via
    :func:`scp_sdk.connect_relay`.

    Args:
        scp: The :class:`scp_sdk.SCP` instance to suspend.

    Raises:
        TransportError: If transport cleanup fails.
    """
    try:
        scp._native.suspend()
    except Exception as exc:  # PyO3 raises ScpTransportError
        raise TransportError(
            f"suspend failed: {exc}",
            code="SCP-TRANS-5001",
        ) from exc


async def resume(scp: SCP) -> None:
    """Resume a suspended bridge instance.

    Clears the suspended flag so bridge operations can proceed.  As of
    Phase 4 PR 3 (#1678) the bridge also reconnects the transport to
    every relay URL the instance was subscribed to at suspend time —
    callers no longer need to re-invoke :func:`scp_sdk.connect_relay`
    manually.

    Delegates to the SDK wrapper's :meth:`SCP.resume` which already runs
    the blocking FFI call in :func:`asyncio.to_thread`.

    Args:
        scp: The :class:`scp_sdk.SCP` instance to resume.

    Raises:
        ContextError: If the bridge has been permanently shut down
            (``shutdown_runtime`` was already called).
    """
    try:
        await asyncio.to_thread(scp._native.resume)
    except Exception as exc:  # PyO3 raises ScpContextError
        # The PyO3 bridge emits SCP-CTX-2000 for resume failures (matching
        # the NAPI and UniFFI bridges — see `scp-ffi/src/lib.rs::scp_resume`).
        # Keeping the code hardcoded here makes the SDK contract explicit
        # even when running against an older bridge where the PyO3
        # constructor used the CTX_2001 default.
        raise ContextError(
            f"resume failed: {exc}",
            code="SCP-CTX-2000",
        ) from exc
